Improvements to the Step by step tutorial (#6622)
Particular/docs.particular.net: tutorials/nservicebus-step-by-step/4-publishing-events/tutorial.md

---
title: "NServiceBus Step-by-step: Publishing Events"
reviewed: 2024-05-21
summary: In this 25-30 minute tutorial, you'll learn how to use the Publish/Subscribe pattern in NServiceBus to define, publish, and subscribe to an event.
redirects:
- tutorials/intro-to-nservicebus/4-publishing-events
- tutorials/nservicebus-101/lesson-4
extensions:
- !!tutorial
  nextText: "Next: Retrying errors"
  nextUrl: tutorials/nservicebus-step-by-step/5-retrying-errors
---

> So far in this tutorial, we have only sent **commands**, meaning one-way messages from a sender to a specific receiver. There's another type of message we have yet to cover called an **event**. In many ways events are just like commands. They're simple classes and you deal with them in much the same way. But from an architectural standpoint, commands and events are *polar opposites*. This creates a useful dichotomy. We can take advantage of the properties of events to open up new possibilities in how we design software.

Reviewer: I feel like the original with the em-dash might read better in this case? But that's just preference.

Regardless,

```suggestion
So far in this tutorial, we have only sent **commands**, meaning one-way messages from a sender to a specific receiver. There's another type of message we have yet to cover called an **event**. In many ways, events are just like commands. They're simple classes and you deal with them in much the same way. But from an architectural standpoint, commands and events are *polar opposites*. This creates a useful dichotomy. We can take advantage of the properties of events to open up new possibilities in how we design software.
```

In the next 25-30 minutes, you will learn how the publish/subscribe pattern can help you create more maintainable code. Together, we'll learn to define, publish, and subscribe to an event.

## Events

> An **event** is another type of message that is published to multiple receivers, unlike a command which is sent to only one receiver. Let's take a look at the formal definitions for commands and events:

Reviewer: ```suggestion
Unlike a **command** which is sent to only one receive, an **event** is another type of message that is published to multiple receivers. Let's take a look at the formal definitions for commands and events:
```

snippet: Event

> All the other considerations for command messages apply to events as well. Properties can be simple types, complex types, or collections, it will depend on what the message serializer supports.

Reviewer: ```suggestion
All the other considerations for command messages apply to events as well. Properties can be simple types, complex types, or collections, depending on what the message serializer supports.
```

## Exercise

Now that we've learned about events and the publish/subscribe pattern, let's make use of it in our ordering system. When a user places an order, we're going to publish an `OrderPlaced` event, then subscribe to it from two brand new endpoints: **Billing** and **Shipping**.

> We'll also create a new OrderBilled event that will be published by the **Billing** endpoint once the credit card transaction is complete.

Reviewer: ```suggestion
We'll also create a new `OrderBilled` event that will be published by the **Billing** endpoint once the credit card transaction is complete.
```
